### ISSUE TITLE
Allow apply refactorings to an already parsed module

### ISSUE THREAD

**Issue by jneira** [mpickering/apply-refact#91]
Hi, we added recently hlint support in haskell-language-server, including the apply of refactorings usgins appy-refact, via `applyRefactorings`
The actual integration could be improved if the api would expose a function to apply them to an already parsed module.
Now we have to use a temporary file and ask apply-refact to parse it again:

```haskell
res <- liftIO $ withSystemTempFile (takeFileName fp) $ \temp h -> do
            hClose h
            writeFileUTF8NoNewLineTranslation temp oldContent
            (Right <$> applyRefactorings Nothing commands temp) `catches`
                    [ Handler $ \e -> return (Left (show (e :: IOException)))
                    , Handler $ \e -> return (Left (show (e :: ErrorCall)))
                    ]
```

As we are using [ghc-lib to parse for ghc < 8.10 and the real ghc for 8.10](https://github.com/haskell/haskell-language-server/blob/master/plugins/hls-hlint-plugin/src/Ide/Plugin/Hlint.hs#L168-L196), following hlint, maybe only we can take advantage in the latter path, but i think it still will worth it.

`Refact.Internal.runRefactoring` is exposed but maybe a higher endpoint, closer to `applyRefactorings` would be handy for downstream packages. 

For ghc < 8.10 we will continue to need delegate the parsing to apply-refact. In that case it would be handy be able to inform the parser option, to, for example, pass custom language extensions not included in the module as pragmas. 
Let me know if you think that should be in another issue.

//cc @ndmitchell @alanz

**Comment by zliu41** [comment]
Sounds good, happy to do this.
Does something like `applyRefactorings`, but replacing the `FilePath` argument with 
```haskell
(Language.Haskell.GHC.ExactPrint.Anns, Located (HsModule GhcPs))
```
work for you?

> it would be handy be able to inform the parser option, to, for example, pass custom language extensions not included in the module as pragmas

Totally agree. `applyRefactorings` should support passing language extensions.

**Comment by jneira** [comment]
Many thanks for take care 😃 

> Does something like applyRefactorings, but replacing the FilePath argument with

Sounds good, we already have `ParsedSource(==Located (HsModule GhcPs))` and only will nees `Language.Haskell.GHC.ExactPrint.Anns` (i guess we can get it from ghc `ApiAnns`??)

**Comment by zliu41** [comment]
Correct, `Anns` can be obtained from `ApiAnns` via [postParseTransform](https://hackage.haskell.org/package/ghc-exactprint-0.6.3.3/docs/Language-Haskell-GHC-ExactPrint-Parsers.html#v:postParseTransform).

**Comment by jneira** [comment]
> For ghc < 8.10 we will continue to need delegate the parsing to apply-refact. In that case it would be handy be able to inform the parser option, to, for example, pass custom language extensions not included in the module as pragmas.

It is done in #98

**Comment by zliu41** [comment]
@jneira It just occurred to me that for this to work, your `Located (HsModule GhcPs)` must be from GHC, not ghc-lib, since that's what ghc-exactprint requires. It seems you are using ghc-lib at least for GHC < 8.10, and if so, this is not going to work for you until ghc-exactprint is migrated to ghc-lib (https://github.com/alanz/ghc-exactprint/issues/93).

I'll fix the other bugs you reported in the next few days but this is the best that can be done at the moment.

I did plan to work on https://github.com/alanz/ghc-exactprint/issues/93 once I get more bandwidth but the pandemic has made it difficult.

**Comment by jneira** [comment]
thanks for your great work
I was assuming that i was going to use `ParsedModule` only for ghc-8.10, at least initially, so it would be great to have it, even without ghc-lib support

**Comment by zliu41** [comment]
OK, sounds good. I'll put it in then.

**Comment by zliu41** [comment]
Merged #100
